src/react/Reconciler.js

```diff
@@ -0,0 +1,330 @@
+    // as we don't support class components yet, throw a fatal error
+    throw new ExpectedBailOut("class components not yet supported");
+  } else {
+    // otherwise it's a functional component, where the second paramater of the function is "context" (if it exists)
+    if (componentType.$FormalParameters.length > 1) {
+      let secondParam = componentType.$FormalParameters[1];
+      if (t.isIdentifier(secondParam)) {
+        contextName = ((secondParam: any): BabelNodeIdentifier).name;
+      }
+      let contextTypeAnnotation = secondParam.typeAnnotation !== undefined && secondParam.typeAnnotation;
+      // we expect that if there's a context param, it should always have Flow annotations
+      if (!contextTypeAnnotation) {
+        throw new ExpectedBailOut(`root component missing Flow type annotations for the "context" paramater`);
+      }
+      contextTypes = flowAnnotationToObjectTypeTemplate(contextTypeAnnotation);
+    }
+  }
+  return createAbstractObject(realm, contextName, contextTypes);
+}
+
+class Reconciler {
+  constructor(
+    realm: Realm,
+    moduleTracer: ModuleTracer,
+    statistics: ReactStatistics,
+    reactSerializerState: ReactSerializerState
+  ) {
+    this.realm = realm;
+    this.moduleTracer = moduleTracer;
+    this.statistics = statistics;
+    this.reactSerializerState = reactSerializerState;
+  }
+
+  realm: Realm;
+  moduleTracer: ModuleTracer;
+  statistics: ReactStatistics;
+  reactSerializerState: ReactSerializerState;
+
+  render(componentType: ECMAScriptSourceFunctionValue): Effects {
+    return this.realm.wrapInGlobalEnv(() =>
+      this.realm.evaluateForEffects(() => {
+        // initialProps and initialContext are created from Flow types from:
+        // - if a functional component, the 1st and 2nd paramater of function
+        // - if a class component, use this.props and this.context
+        // if there are no Flow types for props or context, we will throw a
+        // FatalError, unless it's a functional component that has no paramater
+        // i.e let MyComponent = () => <div>Hello world</div>
+        try {
+          let initialProps = getInitialProps(this.realm, componentType);
+          let initialContext = getInitialContext(this.realm, componentType);
+          let { result } = this._renderAsDeepAsPossible(
+            componentType,
+            initialProps,
+            initialContext,
+            BranchStatus.NO_BRANCH
+          );
+          this.statistics.optimizedTrees++;
+          return result;
+        } catch (error) {
+          // if there was a bail-out on the root component in this reconcilation process, then this
+          // should be an invariant as the user has explicitly asked for this component to get folded
+          let message;
+          if (error instanceof ExpectedBailOut) {
+            message = "bail-out: " + error.message;
+          } else {
+            message = "evaluation bail-out";
```

Comment:
This would catch and cover up all kinds of internal errors, including invariant violations, or SyntaxError/ReferenceError/TypeError in prepack code. Please make this more selective, and rethrow everything else.